File: apps/desktop/src/lib/project-map-scene.ts
import { invoke as tauriInvoke } from '@tauri-apps/api/core';
import { CURRENT_PROTOCOL_VERSION, type InvokeCommand } from './health';
import type { ModuleDependencyRelationV1 } from './module-dependency-graph';

const STABLE_ID_PATTERN = /^[0-9a-f]{64}$/;
const COUNT_PATTERN = /^(?:0|[1-9][0-9]{0,19})$/;
const MAX_U64 = BigInt('18446744073709551615');
const MAX_OVERVIEW_MODULES = 64;
const MAX_FOCUS_MODULES = 32;
const MAX_RELATIONS = 128;
const MAX_INSPECTED_EDGES = BigInt(4_096);
const MAX_DISPLAY_CHARACTERS = 256;

export interface ProjectMapSceneQueryV1 {
  focusModuleId: string | null;
}

export interface QueryProjectMapSceneRequestV1 extends ProjectMapSceneQueryV1 {
  protocolVersion: typeof CURRENT_PROTOCOL_VERSION;
}

export type ProjectMapMappingStatusV1 = 'current' | 'stale' | 'needsReview' | 'unmapped';
export type ProjectMapSceneModuleKindV1 = 'manifestBoundary' | 'pathBoundary';

export interface ProjectMapSceneCardBindingV1 {
  cardId: string;
  sourceIndexRunId: string;
  sourceSnapshotId: string;
}

export interface ProjectMapSceneModuleV1 {
  cardBinding: ProjectMapSceneCardBindingV1 | null;
  cardCoverageBasisPoints: number | null;
  centralSymbolCount: string;
  displayName: string;
  entrypointCount: string;
  fileCount: string;
  kind: ProjectMapSceneModuleKindV1;
  manifestCount: string;
  mappingStatus: ProjectMapMappingStatusV1;
  moduleId: string;
  parentModuleId: string | null;
  rank: number;
  representativeEvidenceId: string | null;
  symbolCount: string;
  testCount: string;
}

export interface ProjectMapSceneRelationV1 {
  evidenceId: string | null;
  observedEvidenceCount: string;
  relation: ModuleDependencyRelationV1;
  sourceModuleId: string;
  targetModuleId: string;
}

export interface ProjectMapSceneV1 {
  focusModuleId: string | null;
  indexRunId: string;
  inspectedEdgeCount: string;
  modules: ProjectMapSceneModuleV1[];
  modulesTruncated: boolean;
  observedRelationGroupCount: string;
  policyVersion: 'v1';
  primaryModuleCount: string;
  relations: ProjectMapSceneRelationV1[];
  relationsTruncated: boolean;
  snapshotId: string;
  sourceEdgesTruncated: boolean;
  unmappedEdgeCount: string;
}

export type ProjectMapSceneResultV1 =
  | { status: 'noProject' }
  | { status: 'noPublishedIndex' }
  | { status: 'projectionUnavailable' }
  | { status: 'focusUnavailable' }
  | { scene: ProjectMapSceneV1; status: 'available' };

export interface ProjectMapSceneResponseV1 {
  protocolVersion: typeof CURRENT_PROTOCOL_VERSION;
  result: ProjectMapSceneResultV1;
}

const invokeThroughTauri: InvokeCommand = (command, arguments_) =>
  tauriInvoke<unknown>(command, arguments_);

export async function queryProjectMapScene(
  query: ProjectMapSceneQueryV1,
  invokeCommand: InvokeCommand = invokeThroughTauri,
): Promise<ProjectMapSceneResponseV1> {
  if (!(query.focusModuleId === null || isStableId(query.focusModuleId))) {
    throw new Error('Project Map scene query does not match V1.');
  }
  const request: QueryProjectMapSceneRequestV1 = {
    focusModuleId: query.focusModuleId,
    protocolVersion: CURRENT_PROTOCOL_VERSION,
  };
  const payload = await invokeCommand('query_project_map_scene', { request });
  const response = parseProjectMapSceneResponseV1(payload);
  if (
    response.result.status === 'available' &&
    response.result.scene.focusModuleId !== query.focusModuleId
  ) {
    throw new Error('Project Map scene response does not match its focus.');
  }
  return response;
}

export function parseProjectMapSceneResponseV1(payload: unknown): ProjectMapSceneResponseV1 {
  if (
    !isRecord(payload) ||
    !hasExactKeys(payload, ['protocolVersion', 'result']) ||
    payload.protocolVersion !== CURRENT_PROTOCOL_VERSION
  ) {
    throw new Error('Project Map scene response does not match V1.');
  }
  return { protocolVersion: CURRENT_PROTOCOL_VERSION, result: parseResult(payload.result) };
}

function parseResult(value: unknown): ProjectMapSceneResultV1 {
  if (!isRecord(value) || typeof value.status !== 'string') return invalidResult();
  for (const status of [
    'noProject',
    'noPublishedIndex',
    'projectionUnavailable',
    'focusUnavailable',
  ] as const) {
    if (value.status === status && hasExactKeys(value, ['status'])) return { status };
  }
  if (value.status === 'available' && hasExactKeys(value, ['scene', 'status'])) {
    return { scene: parseScene(value.scene), status: 'available' };
  }
  return invalidResult();
}

function invalidResult(): never {
  throw new Error('Project Map scene response contains an invalid result.');
}

function parseScene(value: unknown): ProjectMapSceneV1 {
  if (
    !isRecord(value) ||
    !hasExactKeys(value, [
      'focusModuleId',
      'indexRunId',
      'inspectedEdgeCount',
      'modules',
      'modulesTruncated',
      'observedRelationGroupCount',
      'policyVersion',
      'primaryModuleCount',
      'relations',
      'relationsTruncated',
      'snapshotId',
      'sourceEdgesTruncated',
      'unmappedEdgeCount',
    ]) ||
    !isStableId(value.indexRunId) ||
    !isStableId(value.snapshotId) ||
    !(value.focusModuleId === null || isStableId(value.focusModuleId)) ||
    value.policyVersion !== 'v1' ||
    !isCount(value.primaryModuleCount) ||
    !Array.isArray(value.modules) ||
    value.modules.length >
      (value.focusModuleId === null ? MAX_OVERVIEW_MODULES : MAX_FOCUS_MODULES) ||
    typeof value.modulesTruncated !== 'boolean' ||
    !isCount(value.observedRelationGroupCount) ||
    !Array.isArray(value.relations) ||
    value.relations.length > MAX_RELATIONS ||
    typeof value.relationsTruncated !== 'boolean' ||
    !isCount(value.inspectedEdgeCount) ||
    !isCount(value.unmappedEdgeCount) ||
    typeof value.sourceEdgesTruncated !== 'boolean'
  ) {
    throw new Error('Project Map scene response contains an invalid scene.');
  }
  const modules = value.modules.map(parseModule);
  const moduleIds = new Set(modules.map((module) => module.moduleId));
  const primaryCount = BigInt(value.primaryModuleCount);
  if (
    moduleIds.size !== modules.length ||
    modules.some((module, index) => module.rank !== index + 1) ||
    modules.some(
      (module) => module.parentModuleId !== null && !moduleIds.has(module.parentModuleId),
    ) ||
    primaryCount < BigInt(modules.length) ||
    value.modulesTruncated !== primaryCount > BigInt(modules.length) ||
    (value.focusModuleId !== null && !moduleIds.has(value.focusModuleId))
  ) {
    throw new Error('Project Map scene contains contradictory module bounds.');
  }
  const relations = value.relations.map(parseRelation);
  const relationKeys = new Set(
    relations.map(
      (relation) => `${relation.sourceModuleId}:${relation.targetModuleId}:${relation.relation}`,
    ),
  );
  const observedGroups = BigInt(value.observedRelationGroupCount);
  const inspectedEdges = BigInt(value.inspectedEdgeCount);
  const unmappedEdges = BigInt(value.unmappedEdgeCount);
  if (
    relationKeys.size !== relations.length ||
    relations.some(
      (relation) =>
        !moduleIds.has(relation.sourceModuleId) || !moduleIds.has(relation.targetModuleId),
    ) ||
    observedGroups < BigInt(relations.length) ||
    value.relationsTruncated !== observedGroups > BigInt(relations.length) ||
    inspectedEdges > MAX_INSPECTED_EDGES ||
    unmappedEdges > inspectedEdges ||
    (value.sourceEdgesTruncated && inspectedEdges !== MAX_INSPECTED_EDGES)
  ) {
    throw new Error('Project Map scene contains contradictory relation bounds.');
  }
  return {
    focusModuleId: value.focusModuleId,
    indexRunId: value.indexRunId,
    inspectedEdgeCount: value.inspectedEdgeCount,
    modules,
    modulesTruncated: value.modulesTruncated,
    observedRelationGroupCount: value.observedRelationGroupCount,
    policyVersion: 'v1',
    primaryModuleCount: value.primaryModuleCount,
    relations,
    relationsTruncated: value.relationsTruncated,
    snapshotId: value.snapshotId,
    sourceEdgesTruncated: value.sourceEdgesTruncated,
    unmappedEdgeCount: value.unmappedEdgeCount,
  };
}

function parseModule(value: unknown): ProjectMapSceneModuleV1 {
  if (
    !isRecord(value) ||
    !hasExactKeys(value, [
      'cardBinding',
      'cardCoverageBasisPoints',
      'centralSymbolCount',
      'displayName',
      'entrypointCount',
      'fileCount',
      'kind',
      'manifestCount',
      'mappingStatus',
      'moduleId',
      'parentModuleId',
      'rank',
      'representativeEvidenceId',
      'symbolCount',
      'testCount',
    ]) ||
    !isStableId(value.moduleId) ||
    !(value.parentModuleId === null || isStableId(value.parentModuleId)) ||
    (value.kind !== 'manifestBoundary' && value.kind !== 'pathBoundary') ||
    typeof value.displayName !== 'string' ||
    value.displayName.length === 0 ||
    Array.from(value.displayName).length > MAX_DISPLAY_CHARACTERS ||
    containsControl(value.displayName) ||
    !Number.isInteger(value.rank) ||
    typeof value.rank !== 'number' ||
    value.rank < 1 ||
    !isCount(value.manifestCount) ||
    !isCount(value.fileCount) ||
    !isCount(value.symbolCount) ||
    !isCount(value.centralSymbolCount) ||
    !isCount(value.entrypointCount) ||
    !isCount(value.testCount) ||
    !isMappingStatus(value.mappingStatus) ||
    !(value.cardCoverageBasisPoints === null || isBasisPoints(value.cardCoverageBasisPoints)) ||
    !(value.representativeEvidenceId === null || isStableId(value.representativeEvidenceId))
  ) {
    throw new Error('Project Map scene contains an invalid module.');
  }
  const fileCount = BigInt(value.fileCount);
  const symbolCount = BigInt(value.symbolCount);
  if (
    fileCount > symbolCount ||
    BigInt(value.centralSymbolCount) > symbolCount ||
    BigInt(value.entrypointCount) > symbolCount ||
    BigInt(value.testCount) > symbolCount
  ) {
    throw new Error('Project Map scene module counts are contradictory.');
  }
  const cardBinding = value.cardBinding === null ? null : parseCardBinding(value.cardBinding);
  if (
    (value.mappingStatus === 'unmapped') !== (cardBinding === null) ||
    (value.cardCoverageBasisPoints === null) !== (cardBinding === null)
  ) {
    throw new Error('Project Map scene module has contradictory mapping state.');
  }
  return {
    cardBinding,
    cardCoverageBasisPoints: value.cardCoverageBasisPoints,
    centralSymbolCount: value.centralSymbolCount,
    displayName: value.displayName,
    entrypointCount: value.entrypointCount,
    fileCount: value.fileCount,
    kind: value.kind,
    manifestCount: value.manifestCount,
    mappingStatus: value.mappingStatus,
    moduleId: value.moduleId,
    parentModuleId: value.parentModuleId,
    rank: value.rank,
    representativeEvidenceId: value.representativeEvidenceId,
    symbolCount: value.symbolCount,
    testCount: value.testCount,
  };
}

function parseCardBinding(value: unknown): ProjectMapSceneCardBindingV1 {
  if (
    !isRecord(value) ||
    !hasExactKeys(value, ['cardId', 'sourceIndexRunId', 'sourceSnapshotId']) ||
    !isStableId(value.cardId) ||
    !isStableId(value.sourceIndexRunId) ||
    !isStableId(value.sourceSnapshotId)
  ) {
    throw new Error('Project Map scene contains an invalid Card binding.');
  }
  return {
    cardId: value.cardId,
    sourceIndexRunId: value.sourceIndexRunId,
    sourceSnapshotId: value.sourceSnapshotId,
  };
}

function parseRelation(value: unknown): ProjectMapSceneRelationV1 {
  if (
    !isRecord(value) ||
    !hasExactKeys(value, [
      'evidenceId',
      'observedEvidenceCount',
      'relation',
      'sourceModuleId',
      'targetModuleId',
    ]) ||
    !isStableId(value.sourceModuleId) ||
    !isStableId(value.targetModuleId) ||
    value.sourceModuleId === value.targetModuleId ||
    !isRelation(value.relation) ||
    !isPositiveCount(value.observedEvidenceCount) ||
    !(value.evidenceId === null || isStableId(value.evidenceId))
  ) {
    throw new Error('Project Map scene contains an invalid relation.');
  }
  return {
    evidenceId: value.evidenceId,
    observedEvidenceCount: value.observedEvidenceCount,
    relation: value.relation,
    sourceModuleId: value.sourceModuleId,
    targetModuleId: value.targetModuleId,
  };
}

function isRelation(value: unknown): value is ModuleDependencyRelationV1 {
  return (
    typeof value === 'string' &&
    [
      'imports',
      'exports',
      'calls',
      'implements',
      'extends',
      'reads',
      'writes',
      'configures',
      'tests',
      'builds',
      'documents',
    ].includes(value)
  );
}

function isMappingStatus(value: unknown): value is ProjectMapMappingStatusV1 {
  return (
    value === 'current' || value === 'stale' || value === 'needsReview' || value === 'unmapped'
  );
}

function isCount(value: unknown): value is string {
  return typeof value === 'string' && COUNT_PATTERN.test(value) && BigInt(value) <= MAX_U64;
}

function isPositiveCount(value: unknown): value is string {
  return isCount(value) && value !== '0';
}

function isStableId(value: unknown): value is string {
  return typeof value === 'string' && STABLE_ID_PATTERN.test(value);
}

function isBasisPoints(value: unknown): value is number {
  return typeof value === 'number' && Number.isInteger(value) && value >= 0 && value <= 10_000;
}

function containsControl(value: string): boolean {
  return Array.from(value).some((character) => {
    const point = character.codePointAt(0);
    return point !== undefined && (point <= 0x1f || (point >= 0x7f && point <= 0x9f));
  });
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function hasExactKeys(value: Record<string, unknown>, expected: string[]): boolean {
  const actual = Object.keys(value).sort();
  const sortedExpected = [...expected].sort();
  return (
    actual.length === sortedExpected.length &&
    actual.every((key, index) => key === sortedExpected[index])
  );
}

File: crates/a3-protocol/src/project_map_scene.rs
use crate::{ModuleDependencyRelationV1, ProtocolVersion};
use serde::{Deserialize, Serialize};

/// Strict map-scene request without paths, cursors, or caller-controlled limits.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct QueryProjectMapSceneRequestV1 {
    protocol_version: ProtocolVersion,
    focus_module_id: Option<String>,
}

impl QueryProjectMapSceneRequestV1 {
    /// Creates an overview or current-module focus request.
    #[must_use]
    pub const fn new(protocol_version: ProtocolVersion, focus_module_id: Option<String>) -> Self {
        Self {
            protocol_version,
            focus_module_id,
        }
    }

    /// Returns the requested protocol version.
    #[must_use]
    pub const fn protocol_version(&self) -> ProtocolVersion {
        self.protocol_version
    }

    /// Returns the optional untrusted stable module identity.
    #[must_use]
    pub fn focus_module_id(&self) -> Option<&str> {
        self.focus_module_id.as_deref()
    }
}

/// Versioned deterministic atlas scene response.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct ProjectMapSceneResponseV1 {
    protocol_version: ProtocolVersion,
    result: ProjectMapSceneResultV1,
}

impl ProjectMapSceneResponseV1 {
    /// Creates the response used when no project is active.
    #[must_use]
    pub const fn no_project() -> Self {
        Self::with_result(ProjectMapSceneResultV1::NoProject)
    }

    /// Creates the response used before the first atomic publication.
    #[must_use]
    pub const fn no_published_index() -> Self {
        Self::with_result(ProjectMapSceneResultV1::NoPublishedIndex)
    }

    /// Creates the response for publications without deterministic modules.
    #[must_use]
    pub const fn projection_unavailable() -> Self {
        Self::with_result(ProjectMapSceneResultV1::ProjectionUnavailable)
    }

    /// Creates the response for a focus module no longer in the current projection.
    #[must_use]
    pub const fn focus_unavailable() -> Self {
        Self::with_result(ProjectMapSceneResultV1::FocusUnavailable)
    }

    /// Creates one bounded current scene.
    #[must_use]
    pub fn available(scene: ProjectMapSceneV1) -> Self {
        Self::with_result(ProjectMapSceneResultV1::Available {
            scene: Box::new(scene),
        })
    }

    const fn with_result(result: ProjectMapSceneResultV1) -> Self {
        Self {
            protocol_version: ProtocolVersion::CURRENT,
            result,
        }
    }

    /// Returns the mutually exclusive scene result.
    #[must_use]
    pub const fn result(&self) -> &ProjectMapSceneResultV1 {
        &self.result
    }
}

/// Availability state for a project-map scene.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase", tag = "status")]
pub enum ProjectMapSceneResultV1 {
    /// No project is active.
    NoProject,
    /// No index is published.
    NoPublishedIndex,
    /// The publication predates deterministic modules.
    ProjectionUnavailable,
    /// The requested focus is absent or supplementary.
    FocusUnavailable,
    /// One current bounded scene is available.
    Available {
        /// Complete scene bound to one run, snapshot, and policy.
        scene: Box<ProjectMapSceneV1>,
    },
}

/// Deterministic scene-selection policy.
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum ScenePolicyVersionV1 {
    /// Manifest-first overview and evidence-ranked direct focus.
    V1,
}

/// Mapping lifecycle shown independently through text, shape, and color.
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum ProjectMapMappingStatusV1 {
    /// Card is current.
    Current,
    /// Direct evidence invalidated the Card.
    Stale,
    /// A dependency change requires review.
    NeedsReview,
    /// No verified Card exists.
    Unmapped,
}

/// Primary deterministic module boundary kind.
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum ProjectMapSceneModuleKindV1 {
    /// Package or workspace manifest boundary.
    ManifestBoundary,
    /// Deterministic structural path boundary.
    PathBoundary,
}

/// Current Card anchors used only for subsequent typed Inspector reads.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct ProjectMapSceneCardBindingV1 {
    card_id: String,
    source_index_run_id: String,
    source_snapshot_id: String,
}

impl ProjectMapSceneCardBindingV1 {
    /// Creates already validated Card anchors.
    #[must_use]
    pub const fn new(
        card_id: String,
        source_index_run_id: String,
        source_snapshot_id: String,
    ) -> Self {
        Self {
            card_id,
            source_index_run_id,
            source_snapshot_id,
        }
    }
}

/// One visible atlas region with bounded counts and progressive-load hooks.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct ProjectMapSceneModuleV1 {
    module_id: String,
    parent_module_id: Option<String>,
    kind: ProjectMapSceneModuleKindV1,
    display_name: String,
    rank: u16,
    manifest_count: String,
    file_count: String,
    symbol_count: String,
    central_symbol_count: String,
    entrypoint_count: String,
    test_count: String,
    mapping_status: ProjectMapMappingStatusV1,
    card_coverage_basis_points: Option<u16>,
    card_binding: Option<ProjectMapSceneCardBindingV1>,
    representative_evidence_id: Option<String>,
}

impl ProjectMapSceneModuleV1 {
    /// Creates one application-validated module projection.
    #[allow(clippy::too_many_arguments)]
    #[must_use]
    pub const fn new(
        module_id: String,
        parent_module_id: Option<String>,
        kind: ProjectMapSceneModuleKindV1,
        display_name: String,
        rank: u16,
        manifest_count: String,
        file_count: String,
        symbol_count: String,
        central_symbol_count: String,
        entrypoint_count: String,
        test_count: String,
        mapping_status: ProjectMapMappingStatusV1,
        card_coverage_basis_points: Option<u16>,
        card_binding: Option<ProjectMapSceneCardBindingV1>,
        representative_evidence_id: Option<String>,
    ) -> Self {
        Self {
            module_id,
            parent_module_id,
            kind,
            display_name,
            rank,
            manifest_count,
            file_count,
            symbol_count,
            central_symbol_count,
            entrypoint_count,
            test_count,
            mapping_status,
            card_coverage_basis_points,
            card_binding,
            representative_evidence_id,
        }
    }
}

/// One non-authoritative SVG route between two visible regions.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct ProjectMapSceneRelationV1 {
    source_module_id: String,
    target_module_id: String,
    relation: ModuleDependencyRelationV1,
    observed_evidence_count: String,
    evidence_id: Option<String>,
}

impl ProjectMapSceneRelationV1 {
    /// Creates one application-validated grouped route.
    #[must_use]
    pub const fn new(
        source_module_id: String,
        target_module_id: String,
        relation: ModuleDependencyRelationV1,
        observed_evidence_count: String,
        evidence_id: Option<String>,
    ) -> Self {
        Self {
            source_module_id,
            target_module_id,
            relation,
            observed_evidence_count,
            evidence_id,
        }
    }
}

/// Complete bounded atlas scene with explicit omission accounting.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct ProjectMapSceneV1 {
    index_run_id: String,
    snapshot_id: String,
    policy_version: ScenePolicyVersionV1,
    focus_module_id: Option<String>,
    primary_module_count: String,
    modules: Vec<ProjectMapSceneModuleV1>,
    modules_truncated: bool,
    observed_relation_group_count: String,
    relations: Vec<ProjectMapSceneRelationV1>,
    relations_truncated: bool,
    inspected_edge_count: String,
    unmapped_edge_count: String,
    source_edges_truncated: bool,
}

impl ProjectMapSceneV1 {
    /// Creates one application-validated scene projection.
    #[allow(clippy::too_many_arguments)]
    #[must_use]
    pub const fn new(
        index_run_id: String,
        snapshot_id: String,
        policy_version: ScenePolicyVersionV1,
        focus_module_id: Option<String>,
        primary_module_count: String,
        modules: Vec<ProjectMapSceneModuleV1>,
        modules_truncated: bool,
        observed_relation_group_count: String,
        relations: Vec<ProjectMapSceneRelationV1>,
        relations_truncated: bool,
        inspected_edge_count: String,
        unmapped_edge_count: String,
        source_edges_truncated: bool,
    ) -> Self {
        Self {
            index_run_id,
            snapshot_id,
            policy_version,
            focus_module_id,
            primary_module_count,
            modules,
            modules_truncated,
            observed_relation_group_count,
            relations,
            relations_truncated,
            inspected_edge_count,
            unmapped_edge_count,
            source_edges_truncated,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn request_rejects_rendering_limits_paths_and_unknown_fields() {
        let json = r#"{"protocolVersion":1,"focusModuleId":null,"limit":500}"#;
        assert!(serde_json::from_str::<QueryProjectMapSceneRequestV1>(json).is_err());
    }
}
